### PR TITLE
Adding Convenience Fns 

### PR DESCRIPTION
**Pull request by jeffiuliano**
for adding components/connecting components.  These avoid having to write your own try-catch blocks for everything by doing it for you.

One could change the behavior of the `add` and related functions, but this approach is non-breaking and totally backward compatible as it just adds new tools to use. 

Right now all these are grouped together, and then assigned to the relevant classes, but they could also just be defined inside the relevant classes (although this allows a few to share one function).

I update the example script using these new functions too, so you can see how it changes the flow of writing such files.

### CODE REVIEW

**Comment by jeffiuliano** [comment]
Also makes it an installable module

File: padloper/structure.py [hunk 1]
```diff
@@ -4955,6 +4964,86 @@ def from_id(cls, id: int):
...
+        return self
+    except VertexAlreadyAddedError as e:
+        print(e)
+        return self.from_db(self.name)
+
+ComponentType.tryadd = _tryadd
+Component.tryadd = _tryadd
+PropertyType.tryadd = _tryadd
+
+def _tryadd_version(self):
```
Comment (jeffiuliano) [inline padloper/structure.py:4990]: probably a better way to do this so it's not just a copy of the above fn with one line different, but this was fast/easy for now.

**Comment by jeffiuliano** [comment]
Oh, I also added some `__repr__` functions so the classes print something useful when you display them

**Comment by ahincks** [comment]
This is similar to ideas I had. What about, though, if we didn't have two methods, `add()` and`tryadd()`, but rather one method with a optional argument: `def add(self, attributes, strict_add=False)`. If `strict_add=True`, then it will throw an exception if the element already exists in the DB. If it is False, it just prints a warning. In both cases, it returns `self`.

So essentially it would work like the current `add()` if `strict_add=True` (except that it also returns `self`), and it will work like your `tryadd()` if `strict_add=False`.

File: padloper/scripts/init_simple-db.py [hunk 2]
```diff
@@ -1,141 +1,82 @@
 from datetime import datetime
-from structure import *
+from padloper.structure import *
```
Comment (ahincks) [inline padloper/scripts/init_simple-db.py:2]: The way `__init.py__` is set up, I believe you can also just do `from padloper import *`.

**Comment by jeffiuliano** [comment]
Yes, that's definitely a good option if you are willing to change the default behavior of the `add` and related functions.  I'll go with that and then see how using it for the readout chain goes!

**Comment by jeffiuliano** [comment]
Updated with those changes, works at least for the simple-db example code.

**Comment by ahincks** [comment]
Implementation of the `strict` parameter looks good!

I presume the `tryadd()` helpers are no longer needed, in which case, can they be removed?

Instead of simple prints, should we use the `warnings` module or something similar for the warnings? Maybe eventually go with the full-fledged `logging` module? But for now I'd be happy to merge with prints and fix this later, unless you want to tackle it now.

**Comment by jeffiuliano** [comment]
Yes, the  tryadd stuff can definitely go (I forgot to commit that change, pushed now).

You're right that the warnings or logging module is probably the right way show warnings that things are already in the db and otherwise log what happens when things are run.  I'm happy enough with printing for now though, so I would vote to merge and come back to improve that later.

**Review by ahincks — APPROVED** [review]

**Pull request merged**